### PR TITLE
Semantics bug fixes

### PR DESCRIPTION
**Pull request by Contentmaudlin**
Issues addressed: #114 #104 #113 

Elaborating on #113 :

Our namespaces are currently broken, but I'll elaborate on the fixes I did here anyway:

We are allowing structs and namespaces to overlap. However, when structs are being defined, the semantic checker ensures that there are no struct fields that conflicts with a namespace name. This is allowed:

```
ns foo = {
   int x = 1;
}

struct foo { int y; }
```

But this is not:
```
ns foo = {
   int x = 1;
}

struct foo { int x; }
```

And since names are prepended after semant, this won't be a problem for the later stages of our pipeline.

### CODE REVIEW

**Comment by lucas-schuermann** [comment]
re: name spacing allowed, can we just enforce unique names for all top level objects like that?

**Comment by lucas-schuermann** [comment]
i guess the way its implemented is actually kinda nice, but could lead to some garbage code

**Review by lucas-schuermann — APPROVED** [review]
Testing shows that the errors are fixed. Did a quick read through, LGTM. I'm trusting the code was pretty rigorously checked already.

File: src/backend/semant.ml [hunk 1]
```diff
@@ -232,10 +235,20 @@ and check_expr tr_env expr =
...
+                                | Id(l) -> l (* for namespace *) 
+                                | _ -> "") 
           in
+         (* check if namespace *) 
+          let ns_name = fname ^ "_" ^ str in 
+         (* ns variables *) 
+          if VarMap.mem ns_name tr_env.scope
+			    then let x = List.hd (VarMap.find ns_name tr_env.scope) in x.var_type else
+        (* ns function *) 
+        (* TODO:  won't actually work *)
```
Comment (lucas-schuermann) [inline src/backend/semant.ml:247]: ?

Comment (Contentmaudlin) [reply src/backend/semant.ml:247]: See my comment on Slack. Namespaces are currently broken...

**Pull request merged**